src/miral/launch_app.cpp

```diff
@@ -61,65 +101,62 @@ auto miral::launch_app_env(
     mir::optional_value<std::string> const& x11_display,
     miral::AppEnvironment const& app_env) -> pid_t
 {
-    static auto const vars_to_remove = mir_env_variables();
+    Environment application_environment;
 
-    pid_t pid = fork();
-
-    if (pid < 0)
+    for (auto const& [key, value]: app_env)
     {
-        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to fork process"}));
-    }
-
-    if (pid == 0)
-    {
-        for (auto const& var : vars_to_remove)
-        {
-            unsetenv(var.c_str());
-        }
-
-        if (x11_display)
+        if (value)
         {
-            setenv("DISPLAY", x11_display.value().c_str(),  true);   // configure X11 socket
+            application_environment.setenv(key, value.value());
         }
         else
         {
-            unsetenv("DISPLAY");
+            application_environment.unsetenv(key);
         }
+    }
 
-        if (wayland_display)
-        {
-            setenv("WAYLAND_DISPLAY", wayland_display.value().c_str(),  true);   // configure Wayland socket
-        }
-        else
-        {
-            unsetenv("WAYLAND_DISPLAY");
-        }
+    if (wayland_display)
+    {
+        application_environment.setenv("WAYLAND_DISPLAY", wayland_display.value()); // configure Wayland socket
+    }
+    else
+    {
+        application_environment.unsetenv("WAYLAND_DISPLAY");
+    }
 
-        for (auto const& env : app_env)
-        {
-            if (env.second)
-            {
-                setenv(env.first.c_str(), env.second.value().c_str(), true);
-            }
-            else
-            {
-                unsetenv(env.first.c_str());
-            }
-        }
+    if (x11_display)
+    {
+        application_environment.setenv("DISPLAY", x11_display.value());   // configure X11 socket
+    }
+    else
+    {
+        application_environment.unsetenv("DISPLAY");
+    }
+
+    auto const exec_env = application_environment.exec_env();
 
-        std::vector<char const*> exec_args;
+    std::vector<char const*> exec_args;
 
-        for (auto const& arg : app)
-            exec_args.push_back(arg.c_str());
+    for (auto const& arg : app)
+        exec_args.push_back(arg.c_str());
 
-        exec_args.push_back(nullptr);
+    exec_args.push_back(nullptr);
+
+    pid_t pid = fork();
 
+    if (pid < 0)
+    {
+        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to fork process"}));
+    }
+
+    if (pid == 0)
+    {
         mir::log_debug("Restoring sigmask");
         sigset_t all_signals;
         sigfillset(&all_signals);
         pthread_sigmask(SIG_UNBLOCK, &all_signals, nullptr);
 
-        execvp(exec_args[0], const_cast<char*const*>(exec_args.data()));
+        execvpe(exec_args[0], const_cast<char*const*>(exec_args.data()), const_cast<char*const*>(exec_env.data()));
```

Comment:
`execvpe`,sadly, isn't async-signal-safe (presumably because the obvious implementation requires memory allocation?).

I believe the correct approach here is to do `$PATH` handling ourselves; pre-construct all the possible paths that would be searched through, and then in the child calling `execve` on each in turn, annoyingly needing to keep track of error returns.

What a faff!

Reply:
For added bonus giggles, if `$PATH` is unset, we should be looking in [`_CS_PATH`](https://www.gnu.org/software/libc/manual/html_node/String-Parameters.html). Dear lord!

Reply:
(For the record, *none* of the `exec*p*` functions are async-signal-safe)

Reply:
> What a faff!

Indeed. This is something that library implementers could (and should) get right for the rest of us

Reply:
I don't think there's a need for memory allocation in "the obvious implementation": remember _C99_ has VLAs. (See https://codebrowser.dev/glibc/glibc/posix/execvpe.c.html for an example implementation.)

So requiring `execvpe()` to be async-signal-safe is plausible (but outside our authority).